File: app/src/main/java/ch/zuehlke/sbb/reddit/data/source/remote/model/posts/RedditPostElement.java
package ch.zuehlke.sbb.reddit.data.source.remote.model.posts;

/**
 * Created by chsc on 13.11.17.
 */


public abstract class RedditPostElement {


    private RedditPostElement() {
    }



    public static RedditPostElement reference(final String reference) {
        return new ReferenceRedditPostElement(reference);
    }

    public static final class DataRedditPostElement extends RedditPostElement {

        public final String kind = null;
        public final RedditPost data = null;


        private DataRedditPostElement() {
        }

    }

    // This is a special wrapper because we cannot make java.util.String to be a subclass of the RedditPostElement class
    // Additionally, you can add more methods if necessary
    public static final class ReferenceRedditPostElement extends RedditPostElement {

        public final String reference;

        private ReferenceRedditPostElement(final String reference) {
            this.reference = reference;
        }

    }

}

File: app/src/main/java/ch/zuehlke/reddit/data/source/remote/model/posts/RedditPostElement.kt
package ch.zuehlke.reddit.data.source.remote.model.posts

import com.google.gson.annotations.Expose

/**
 * Created by chsc on 13.11.17.
 */


abstract class RedditPostElement private constructor() {


    class DataRedditPostElement : RedditPostElement() {

        @Expose
        val kind: String? = null
        @Expose
        val data: RedditPost? = null

    }

    // This is a special wrapper because we cannot make java.util.String to be a subclass of the RedditPostElement class
    // Additionally, you can add more methods if necessary
    class ReferenceRedditPostElement constructor(@Expose
                                                         val reference: String) : RedditPostElement()

    companion object {


        fun reference(reference: String): RedditPostElement {
            return ReferenceRedditPostElement(reference)
        }
    }

}